plugin/trino-pulsar/src/main/java/io/trino/plugin/pulsar/PulsarConnectorCacheImpl.java

```diff
@@ -0,0 +1,208 @@
+/*
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package io.trino.plugin.pulsar;
+
+import com.google.common.annotations.VisibleForTesting;
+import com.google.common.base.Strings;
+import com.google.common.collect.ImmutableMap;
+import io.airlift.log.Logger;
+import org.apache.pulsar.PulsarVersion;
+import org.apache.pulsar.shade.org.apache.bookkeeper.common.util.OrderedScheduler;
+import org.apache.pulsar.shade.org.apache.bookkeeper.conf.ClientConfiguration;
+import org.apache.pulsar.shade.org.apache.bookkeeper.mledger.LedgerOffloader;
+import org.apache.pulsar.shade.org.apache.bookkeeper.mledger.LedgerOffloaderFactory;
+import org.apache.pulsar.shade.org.apache.bookkeeper.mledger.ManagedLedgerConfig;
+import org.apache.pulsar.shade.org.apache.bookkeeper.mledger.ManagedLedgerFactory;
+import org.apache.pulsar.shade.org.apache.bookkeeper.mledger.ManagedLedgerFactoryConfig;
+import org.apache.pulsar.shade.org.apache.bookkeeper.mledger.impl.ManagedLedgerFactoryImpl;
+import org.apache.pulsar.shade.org.apache.bookkeeper.mledger.impl.NullLedgerOffloader;
+import org.apache.pulsar.shade.org.apache.bookkeeper.mledger.offload.Offloaders;
+import org.apache.pulsar.shade.org.apache.bookkeeper.mledger.offload.OffloadersCache;
+import org.apache.pulsar.shade.org.apache.bookkeeper.stats.StatsProvider;
+import org.apache.pulsar.shade.org.apache.pulsar.common.naming.NamespaceName;
+import org.apache.pulsar.shade.org.apache.pulsar.common.policies.data.OffloadPolicies;
+
+import java.io.IOException;
+import java.util.Map;
+import java.util.Objects;
+import java.util.concurrent.ConcurrentHashMap;
+
+import static com.google.common.base.Preconditions.checkNotNull;
+
+/**
+ * A cache for the Pulsar connector.
+ */
+public class PulsarConnectorCacheImpl
+        implements PulsarConnectorCache
+{
+    private static final Logger log = Logger.get(PulsarConnectorCacheImpl.class);
+
+    @VisibleForTesting
+    static PulsarConnectorCache instance;
```

Comment:
This bad design. It is not going to work if user to have several pulsar catalogs. Currently we have classloader separation that each plugin has its own plugin, but catalogs that are using same connector (same plugin then) will share the same classloader. In same cases we go even further and we can duplicate classloader per each catalog, but still this design is bad.

Please use guice to inject that instance.

Reply:
Sure thing, changed to use Guice to inject it.